symfony/symfony
# [DI] Randomize private services

| Q | A |
| --- | --- |
| Branch? | "master" |
| Bug fix? | yes |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | not yet |
| Fixed tickets | #19680 |
| License | MIT |
| Doc PR | reference to the documentation PR, if any |

Alternative approach compared to #19682 

Thanks to @wouterj for explaining this approach.

Theory of operation;
- compiler pass
  - build a map of `private id` => `random id`
  - remap private definitions to its new id
  - update references and aliases
  - inject id map into container builder (for BC)
- container
  - id's found in the (injected) map are forwarded to the new id + deprecation notice
- todo
  - add tests
  - fix tests due randomization
  - dump the container with orignal id's in generated docs, etc. for convenience.

The deprecation now only happens after compilation(!). This can be seen a design choice, i guess.
